### PR TITLE
Framework: Stop caching the user in local storage

### PR DESCRIPTION
**Pull request by drewblaisdell**
The user is always fresh when the client boots in production because user bootstrapping is enabled, but is always stale in development because the user object comes from local storage before it is fetched. This makes it possible for us to miss production-only bugs or spend time fixing development-only bugs.

For example, we have code that determines if we need to refetch `sites-list` before redirecting to the sitepicker if the user accesses a site that is missing from `sites-list` [here](https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/controller.js#L285). We check if `sites-list` is missing a site by comparing its length to the `visible_site_count` property of the user. This is reliable in production, as the user object is provided in the page, but breaks in development in some cases (e.g. if the site was created on a different browser, computer, or outside of Calypso during a PayPal Express checkout) as both the user object and `sites-list` will be stale.

This PR updates `lib/user` to stop storing the user in local storage, which decreases the disparity of behavior between development/production. The downside to this is that the boot is slightly longer in development as the client doesn't boot [until the user is fetched](https://github.com/Automattic/wp-calypso/blob/master/client/boot/index.js#L432), but only slightly and worth the improved resistance to environment-specific bugs in my opinion.

**Testing**
Almost every page in Calypso relies on `lib/user`. You can assert that this does not break anything in development by loading any page under My Sites or the Reader.

cc @gwwar @blowery who also discussed this on Slack

### CODE REVIEW

**Comment by blowery** [comment]
One other thing to test here: how this interacts with lib/user/support-user-interop. Should have no / little impact, but worth poking at.

**Comment by drewblaisdell** [comment]
Thanks for pointing that out. I was able to `su` into another user on this branch properly.

**Comment by gwwar** [comment]
Let's also cc @rralian and @jordwest for help testing. We'll want to verify both local dev and the docker image for regressions.

At the moment it looks like there are a few unit tests that need to be updated?

**Comment by rralian** [comment]
The idea makes sense and sounds reasonable. I spot-tested a bit and could find no issues.

**Comment by drewblaisdell** [comment]
Thanks for looking at this @gwwar / @rralian.

> At the moment it looks like there are a few unit tests that need to be updated?

Apologies for missing this. My initial diff changed the behavior of `lib/user` slightly such that `user.data` was undefined instead of explicitly set to `false`, as I simply removed `this.data = store.get( 'wpcom_user' ) || false;` in the condition where user bootstrapping is disabled. This was causing the tests to fail (because `( false ).foo === undefined` but `( undefined ).foo` throws a type error). I updated the commit such that the `data` property of the user is initially `false` to preserve the current behavior.

**Comment by jordwest** [comment]
🎉  This is great!

If anything, this should simplify support user as avoiding the various caching was one of the roadblocks to having a clean user state post-switch (eg, the `this.data = false` removed in this PR). I don't see any reason it should negatively affect support user.

Tested switching users with different locales (an earlier sticking point in support user). Tested purchases, billing history, stats, and that /help/contact shows the correct support for the user's account. All worked as expected.

File: client/lib/user/user.js [hunk 1]
```diff
@@ -48,12 +47,11 @@ User.prototype.initialize = function() {
 	debug( 'Initializing User' );
 	this.fetching = false;
 	this.initialized = false;
+	this.data = false;
 
 	this.on( 'change', this.checkVerification.bind( this ) );
 
 	if ( isSupportUserSession() ) {
-		this.data = false;
-
```
Comment (jordwest) [inline client/lib/user/user.js]: 🎉

**Comment by mtias** [comment]
How does this interact with oauth setups (desktop and development versions).

**Comment by drewblaisdell** [comment]
@mtias There shouldn't be any interaction as the client doesn't boot [until the user is fetched](https://github.com/Automattic/wp-calypso/blob/master/client/boot/index.js#L432), regardless of the OAuth setting.

**Comment by drewblaisdell** [comment]
Thanks everyone.

**Comment by drewblaisdell** [comment]
This PR had to be reverted because it disabled the client boot when bootstrapping is enabled and the user is logged out (e.g. `/start` in production). I need to investigate the exact mechanism for this, but I suspect that the client boot function was never triggered, as there were no errors in the console, writes to the template, or anything else we would expect to happen as soon as that happens.

In the future, I'm going to check any user-specific changes against a logged out route (e.g. `/start`) before deploying. @alisterscott updated some internal docs to this effect as well.

**Pull request merged**